File: vcone.py
import numpy as np


class ViewingCone(object):
    """
    A class representing viewing cones obtained by LINE21.
    """

    def __init__(self, name):

        if name.lower() == 'km11' or name.lower() == 'tofor':
            file_path = 'vc_data/TOFOR/TOFOR.vc'
        elif name.lower() == 'ab':
            file_path = 'vc_data/AB/AB.vc'
        else:
            file_path = name

        self.file = file_path

        VC = np.loadtxt(file_path)

        self.X   = VC[:,0]
        self.Y   = VC[:,1]
        self.Z   = VC[:,2]
        self.C   = VC[:,3]
        self.V   = VC[:,4]
        self.UX  = VC[:,5]
        self.UY  = VC[:,6]
        self.UZ  = VC[:,7]
        self.R   = VC[:,8]
        self.PHI = VC[:,9]
        if VC.shape[1]==11:
            self.OMEGA = VC[:,10]
        else:
            self.OMEGA = 4*np.pi * self.C / self.V

        self.Nvoxels = len(self.C)
        self.Vtot = np.sum(self.V)

        # Transform detector direction vectors to cylindrical coordinates
        self.UR   = self.UX*np.cos(self.PHI) + self.UY*np.sin(self.PHI)
        self.Uphi = -self.UX*np.sin(self.PHI) + self.UY*np.cos(self.PHI)

        # Collect cylindrical direction vectors
        self.U = np.vstack((self.UR, self.Uphi, self.UZ))

        # Reconstruct the grid used by LINE21
        self.Rvals = np.unique(self.R)
        self.Zvals = np.unique(self.Z)
        self.PHIvals = np.unique(self.PHI)
        self.NR = len(self.Rvals)
        self.NZ = len(self.Zvals)
        self.NP = self.NR * self.NZ    # number of poloidal grid points
        self.NPHI = len(self.PHIvals)   # number of toroidal grid points

        self.dR = self.Rvals[1] - self.Rvals[0]    # assumes uniform grid spacing
        self.dZ = self.Zvals[1] - self.Zvals[0]
        self.dPHI = self.PHIvals[1] - self.PHIvals[0]

        # Map each voxel to the corrresponding poloidal grid point
        self.IR =  np.round((self.R - self.Rvals[0]) / self.dR).astype(int)
        self.IZ =  np.round((self.Z - self.Zvals[0]) / self.dZ).astype(int)
        self.IP = self.IR + self.IZ*self.NR

        self.IPvals = np.unique(self.IP)

        # Map each poloidal grid point to the corresponding voxels
        IVOX = np.empty(self.NP, dtype='object')
        for i in range(self.NP):
            IVOX[i] = []

        for ivox, ip in enumerate(self.IP):
            IVOX[ip].append(ivox)

        self.IVOX = IVOX

        # Also count the number of voxels at each poloidal grid point
        NVOX = np.zeros(self.NP, dtype='int')

        for i in range(self.NP):
            NVOX[i] = len(IVOX[i])

        self.NVOX = NVOX

        # Generate poloidal projection of the viewing cone
        IRP = self.IPvals%self.NR
        IZP = self.IPvals//self.NR
        self.RP = self.Rvals[IRP]
        self.ZP = self.Zvals[IZP]

        CP  = np.zeros(len(self.IPvals))

        for i,ip in enumerate(self.IPvals):
            # Total weight of all voxels at the given poloidal location
            CP[i]  = np.sum(self.C[IVOX[ip]])

        self.CP = CP


    def __repr__(self):
        return 'Viewing cone: {}'.format(self.file)


    # Class methods
    def get_poloidal_index(self, r, z):
        """ Check which poloidal bin that we are in.
        Requires that the poloidal grid spacing is uniform."""
        r = np.atleast_1d(r)
        z = np.atleast_1d(z)

        ir = np.floor( ( r - self.Rvals[0] + self.dR/2.0) / self.dR ).astype(int)
        iz = np.floor( ( z - self.Zvals[0] + self.dZ/2.0) / self.dZ ).astype(int)

        ip = ir + iz*self.NR

        ip[ir<0]         = -1
        ip[ir>=self.NR] = -1
        ip[iz<0]         = -1
        ip[iz>=self.NZ] = -1

        outside = np.logical_not(np.in1d(ip, self.IP))
        ip[outside] = -1

        return ip


    def is_inside(self, r, z):
        """ Check whether the given points are inside the viewing cone. """

        ip = self.get_poloidal_index(r, z)

        is_inside = np.in1d(ip, self.IP)

        return is_inside


    def get_voxels(self, r, z):
        """ Return all voxels that include a given point in the poloidal plane. """

        ip = self.get_poloidal_index(r, z)

        return np.where(ip==-1, -1, self.IVOX[ip])


    def map_points_voxels(self, r, z):
        """
        Map given (r,z) points to each voxel that encloses them, under the
        assumption of toroidal symmetry. 
        
        Returns
        -------
        i_voxels: array of voxel indices. Each element correponds to 
            one point in the voxel with that index.

        i_points: array of point indices (same length as i_voxels). 
            For instance, i_point[j] tells us that that point is
            inside the voxel with index i_voxels[j].
            (A point with given r,z coordinates might be inside 
            multiple voxels, since toroidal symmetry is assumed).

        (DISABLED) n_points: array that, for each point-voxel pair, holds the 
            number of points in that voxel.
        """

        ip = self.get_poloidal_index(r, z)
        i_all_points = np.arange(len(ip))
        
        inside = np.in1d(ip, self.IP)
        ip = ip[inside]

        if len(ip) == 0:
            # No points inside any voxel
            return np.array([]), np.array([])

        i_points_inside = i_all_points[inside]

        ivox = self.IVOX[ip]
        nvox = self.NVOX[ip]
        
        #ip_inverse, n_counts = np.unique(ip, return_inverse=True, return_counts=True)[1:]
        #n_counts = n_counts[ip_inverse]

        i_voxels = np.concatenate(ivox)
        i_points = np.zeros_like(i_voxels)
        #n_points = np.zeros_like(i_voxels)

        i0 = 0
        for i,n in enumerate(nvox):
            i1 = i0 + n
            i_points[i0:i1] = i_points_inside[i] 
            #n_points[i0:i1] = n_counts[i]
            i0 = i1

        return i_voxels, i_points #, n_points

File: vcone.jl
########################################## vcone.jl ##################################################
# This script contains structures and functions to be able to load and represent a viewing cone 
# (aka line-of-sight, aka sightline) stored in a file produced by the LINE21 code (or similar).
# It is a direct Julia re-implementation of the vcone.py file in the DRESS code (J. Eriksson et al, 
# Computer Physics Communications 199 (2016) 40–46).
#
# THIS SCRIPT IS FUNCTIONALLY DONE. ONLY DOCUMENTATION REMAINS TO BE FINALIZED.
#
# Written by Henrik Järleblad. Last maintained 2024-10-30
######################################################################################################

using DelimitedFiles
using LinearAlgebra

"""
    ViewingCone

STRUCTURE DESCRIPTION TO BE WRITTEN HERE.
"""
struct ViewingCone
    file::String
    X::Vector{Float64}
    Y::Vector{Float64}
    Z::Vector{Float64}
    C::Vector{Float64}
    V::Vector{Float64}
    UX::Vector{Float64}
    UY::Vector{Float64}
    UZ::Vector{Float64}
    R::Vector{Float64}
    PHI::Vector{Float64}
    OMEGA::Vector{Float64}
    Nvoxels::Int64
    Vtot::Float64
    UR::Vector{Float64}
    Uphi::Vector{Float64}
    U::Matrix{Float64}
    Rvals::Vector{Float64}
    Zvals::Vector{Float64}
    PHIvals::Vector{Float64}
    NR::Int64
    NZ::Int64
    NP::Int64
    NPHI::Int64
    dR::Float64
    dZ::Float64
    dPHI::Float64
    IR::Vector{Int64}
    IZ::Vector{Int64}
    IP::Vector{Int64}
    IPvals::Vector{Int64}
    IVOX::Vector{Vector{Int64}}
    NVOX::Vector{Int64}
    RP::Vector{Float64}
    ZP::Vector{Float64}
    CP::Vector{Float64}
end

"""
    ViewingCone(name)

FUNCTION DESCRIPTION TO BE WRITTEN HERE
"""
function ViewingCone(name::String)
    # Determine file path based on name
    file_path = (lowercase(name) == "km11" || lowercase(name) == "tofor") ? "vc_data/TOFOR/TOFOR.vc" :
                (lowercase(name) == "ab" ? "vc_data/AB/AB.vc" : name)

    # Load data from file
    VC = readdlm(file_path)

    # Assign columns to respective variables
    X, Y, Z = VC[:, 1], VC[:, 2], VC[:, 3]
    C, V = VC[:, 4], VC[:, 5]
    UX, UY, UZ = VC[:, 6], VC[:, 7], VC[:, 8]
    R, PHI = VC[:, 9], VC[:, 10]

    # Calculate OMEGA if not present
    OMEGA = size(VC, 2) == 11 ? VC[:, 11] : ((4*π .*C) ./ V)

    # Basic attributes
    Nvoxels = length(C)
    Vtot = sum(V)

    # Convert detector direction vectors to cylindrical coordinates
    UR = UX .* cos.(PHI) .+ UY .* sin.(PHI)
    Uphi = -UX .* sin.(PHI) .+ UY .* cos.(PHI)
    U = hcat(UR, Uphi, UZ)'

    # Reconstruct grid used by LINE21 and unique values
    Rvals, Zvals, PHIvals = sort(unique(R)), sort(unique(Z)), sort(unique(PHI))
    NR, NZ, NPHI = length(Rvals), length(Zvals), length(PHIvals)
    NP = NR * NZ # Number of poloidal grid points

    # Grid spacing (assumes uniform)
    dR, dZ, dPHI = Rvals[2] - Rvals[1], Zvals[2] - Zvals[1], PHIvals[2] - PHIvals[1]

    # Map each voxel to the corrresponding poloidal grid point
    IR = round.(Int64, (R .- Rvals[1]) ./ dR) # Indices starting at 0, due to easy math operations
    IZ = round.(Int64, (Z .- Zvals[1]) ./ dZ) # Indices starting at 0, due to easy math operations
    IP = IR + NR .*IZ # Indices starting at 0, due to easy math operations
    IPvals = sort(unique(IP)) # Indices starting at 0, due to easy math operations

    # Map each poloidal grid point to the corresponding voxels
    IVOX = [Int64[] for _ in 1:NP] # A Vector of (empty) Vectors of Int64 elements
    for (ivox, ip) in enumerate(IP)
        push!(IVOX[ip + 1], ivox)
    end

    # Count number of voxels at each poloidal grid point
    NVOX = [length(IVOX[i]) for i in 1:NP]

    # Generate poloidal projection of the viewing cone
    IRP, IZP = IPvals .% NR, div.(IPvals, NR) # The remainder is the R index. The quotient is the z index
    RP, ZP = Rvals[IRP .+ 1], Zvals[IZP .+ 1]
    CP = [sum(C[IVOX[ip + 1]]) for ip in IPvals] # Total weight of all voxels at the given poloidal location

    # Correct for Julia array index convention (start at index 1 instead of 0)
    IR .+= 1
    IZ .+= 1
    IP .+= 1
    IPvals .+= 1

    return ViewingCone(file_path, X, Y, Z, C, V, UX, UY, UZ, R, PHI, OMEGA, Nvoxels, Vtot, UR, Uphi, U, Rvals, Zvals, PHIvals,
                       NR, NZ, NP, NPHI, dR, dZ, dPHI, IR, IZ, IP, IPvals, IVOX, NVOX, RP, ZP, CP)
end

"""
get_poloidal_index(vc,r,z)

Method to check poloidal index. 
Check which poloidal bin the input (R,z) is in.
Requires that the poloidal grid spacing is uniform.

WRITE EXPLANATION FOR INPUTS AND OUTPUT
"""
function get_poloidal_index(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}) where {T<:AbstractArray}
    R, z = R |> vcat, z |> vcat  # Convert scalars to 1-element Vectors
    ir = Int64.(floor.((R .- vc.Rvals[1] .+ (vc.dR / 2)) ./ vc.dR)) # Indices starting at 0, due to easy math operations
    iz = Int64.(floor.((z .- vc.Zvals[1] .+ (vc.dZ / 2)) ./ vc.dZ)) # Indices starting at 0, due to easy math operations
    ip = ir .+ vc.NR .*iz # Indices starting at 0, due to easy math operations

    # Mark out-of-bounds indices
    ip[ir .< 0 .|| ir .>= vc.NR .|| iz .< 0 .|| iz .>= vc.NZ] .= -1 # Set all poloidal indices with (R,z) points outside the grid to -1
    ip = map(x-> x==-1 ? x : x+1, ip) # Return to using 1 to refer to index of first element in array, but keep -1 elements
    outside = .!(in.(ip, Ref(vc.IP))) # If there somehow were indices created outside of the poloidal index range...
    ip[outside] .= -1 # ...set them to -1 as well
    return ip
end

"""
is_inside(vc,R,z)

Method to check if points are inside the viewing cone

WRITE EXPLANATION FOR INPUTS AND OUTPUT
"""
function is_inside(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}) where {T<:AbstractArray}
    ip = get_poloidal_index(vc, R, z)
    return in.(ip, Ref(vc.IP)) # Return an array where 1 means inside the vc, 0 means outside
end

"""
get_voxels(vc,R,z)

Method to get all voxels including a point in the poloidal plane

WRITE EXPLANATION FOR INPUTS AND OUTPUT
"""
function get_voxels(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}) where {T<:AbstractArray}
    ip = get_poloidal_index(vc, R, z)
    return map(x-> x == -1 ? -1 : vc.IVOX[x],ip)
end

"""

map_points_voxels(vc,R,z)

Map given (R,z) points to each voxel that encloses them, under the
assumption of toroidal symmetry. 

Inputs
-------
WRITE EXPLANATION FOR INPUTS

Outputs
-------
i_voxels: array of voxel indices. Each element correponds to 
    one point in the voxel with that index.

i_points: array of point indices (same length as i_voxels). 
    For instance, i_point[j] tells us that that point is
    inside the voxel with index i_voxels[j].
    (A point with given R,z coordinates might be inside 
    multiple voxels, since toroidal symmetry is assumed).
"""
function map_points_voxels(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}) where {T<:AbstractArray}
    ip = get_poloidal_index(vc, R, z)
    i_all_points = collect(1:length(ip))

    # Filter points inside the viewing cone
    inside = findall(x-> x != -1, ip)
    ip, i_points_inside = ip[inside], i_all_points[inside]

    if isempty(ip)
        # No points inside any voxel
        return Int64[], Int64[]
    end

    # Map points to voxel indices
    ivox = vc.IVOX[ip]
    nvox = vc.NVOX[ip]
    
    i_voxels = vcat(ivox) # Reshape all voxel 1-element arrays into one long vector
    i_points = similar(i_voxels) # Pre-allocate i_points

    # Assign points to voxel indices
    i0 = 1
    for (i, n) in enumerate(nvox)
        i1 = i0 + n - 1
        i_points[i0:i1] .= i_points_inside[i]
        i0 = i1 + 1
    end

    return i_voxels, i_points
end
